### ISSUE TITLE
Reproducing sota results for DM-large model on MNIST dataset

### ISSUE THREAD

**Issue by zbh2047** [huanzhang12/CROWN-IBP#3]
Hi,

Thanks so much for sharing the pytorch version CROWN-IBP code. I am really excited about the excellent result (7.02 certified accuracy) on MNIST dataset. However, after a simple try, I have some difficulty in reproducing the results. I wonder if there's something wrong in my experimental setup.

I try this command:
`python train.py --config config/mnist_dm-large_0.4.json`
to use CROWN-IBP, the largest model, and test eps=0.3. However, after 200 epochs, the final outputs are the following:
```
Epoch 199, learning rate [5e-06], epsilon 0.4 - 0.4
[199:   0]: eps 0.400000  Time 0.102 (0.102)  Total Loss 0.2169 (0.2169)  L1 Loss 0.0000 (0.0000)  CE 0.0364 (0.0364)  RCE 0.3975 (0.3975)  Err 0.0156 (0.0156)  Rob Err 0.1055 (0.1055)  Uns 2312.7 (2312.7)  Dead 161882.3 (161882.3)  Alive 11857.7 (11857.7)  Tightness 2312.68750 (2312.68750)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  R 156.462  beta 0.000 (0.000)  kappa 0.500 (0.500)
[199:  50]: eps 0.400000  Time 0.068 (0.070)  Total Loss 0.1610 (0.1710)  L1 Loss 0.0000 (0.0000)  CE 0.0423 (0.0300)  RCE 0.2797 (0.3121)  Err 0.0117 (0.0093)  Rob Err 0.0938 (0.0950)  Uns 2163.9 (2286.2)  Dead 161821.3 (161883.4)  Alive 11882.7 (11852.9)  Tightness 2163.91016 (2286.21722)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  R 159.006  beta 0.000 (0.000)  kappa 0.500 (0.500)
[199: 100]: eps 0.400000  Time 0.072 (0.070)  Total Loss 0.2200 (0.1738)  L1 Loss 0.0000 (0.0000)  CE 0.0582 (0.0307)  RCE 0.3819 (0.3168)  Err 0.0195 (0.0094)  Rob Err 0.0938 (0.0956)  Uns 2254.2 (2246.9)  Dead 161849.0 (161882.8)  Alive 11846.2 (11853.7)  Tightness 2254.18359 (2246.88065)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  R 132.638  beta 0.000 (0.000)  kappa 0.500 (0.500)
[199: 150]: eps 0.400000  Time 0.068 (0.070)  Total Loss 0.2239 (0.1716)  L1 Loss 0.0000 (0.0000)  CE 0.0388 (0.0307)  RCE 0.4090 (0.3126)  Err 0.0156 (0.0091)  Rob Err 0.1172 (0.0940)  Uns 2321.2 (2255.6)  Dead 161854.6 (161883.1)  Alive 11836.2 (11853.0)  Tightness 2321.19141 (2255.60943)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  R 143.678  beta 0.000 (0.000)  kappa 0.500 (0.500)
[199: 200]: eps 0.400000  Time 0.072 (0.070)  Total Loss 0.1639 (0.1730)  L1 Loss 0.0000 (0.0000)  CE 0.0449 (0.0311)  RCE 0.2828 (0.3148)  Err 0.0117 (0.0094)  Rob Err 0.0977 (0.0947)  Uns 2268.5 (2267.8)  Dead 161935.2 (161882.4)  Alive 11859.1 (11854.1)  Tightness 2268.51562 (2267.80655)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  R 141.274  beta 0.000 (0.000)  kappa 0.500 (0.500)
[FINAL RESULT epoch:199 eps:0.4000]: Time 0.067 (0.071)  Total Loss 0.1554 (0.1728)  L1 Loss 0.0000 (0.0000)  CE 0.0183 (0.0307)  RCE 0.2925 (0.3149)  Uns 2321.958 (2266.811)  Dead 161800.2 (161883.5)  Alive 11835.8 (11854.5)  Tight 2321.95833 (2266.81117)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  Err 0.0000 (0.0092)  Rob Err 0.1354 (0.0945)  R 134.250  beta 0.000 (0.000)  kappa 0.500 (0.500)

layer 0 norm 1.8666208982467651
layer 2 norm 16.517566680908203
layer 4 norm 40.252685546875
layer 6 norm 159.90121459960938
layer 8 norm 115.69601440429688
layer 11 norm 655.5496826171875
layer 13 norm 56.66131591796875
Epoch time: 17.0459, Total time: 4751.4625
Evaluating...
[FINAL RESULT epoch:199 eps:0.4000]: Time 0.041 (0.040)  Total Loss 0.1400 (0.4073)  L1 Loss 0.0000 (0.0000)  CE 0.0065 (0.0641)  RCE 0.2734 (0.7504)  Uns 2016.938 (2200.269)  Dead 161932.7 (161853.5)  Alive 12006.9 (11892.8)  Tight 2016.93750 (2200.26890)  Bias 0.00000 (0.00000)  Diff 0.00000 (0.00000)  Err 0.0000 (0.0177)  Rob Err 0.0625 (0.1319)  R 83.985  beta 0.000 (0.000)  kappa 0.500 (0.500)

layer 0 norm 1.8666208982467651
layer 2 norm 16.517566680908203
layer 4 norm 40.252685546875
layer 6 norm 159.90121459960938
layer 8 norm 115.69601440429688
layer 11 norm 655.5496826171875
layer 13 norm 56.66131591796875
saving to ./mnist_dm-large_0.4/IBP_large.pth
Total Time: 4751.4625
Model IBP_large best err 0.1313, clean err 0.0178

```

It seems that the robust error is large. Could you please advise me on how to correctly reproduce your results? Thank you very much!

**Comment by zbh2047** [comment]
I understand that this error may be under eps=0.4. I should use eval.py to get the certified error on test set. I am really sorry that I get the wrong meaning of the output.
